igvm/hypervisor_ranking.py

```diff
@@ -0,0 +1,78 @@
+"""igvm - Hypervisor ranking
+
+Copyright (c) 2018, InnoGames GmbH
+"""
+
+from functools import cmp_to_key
+from logging import getLogger
+from os import environ
+
+from adminapi.dataset import Query
+from adminapi.filters import Any
+
+from igvm.settings import HYPERVISOR_ATTRIBUTES, HYPERVISOR_PREFERENCES
+from igvm.hypervisor import Hypervisor
+
+log = getLogger(__name__)
+
+
+class HypervisorRanking(object):
+    """Encapsulate hypervisors for lazy comparison of their rankings"""
+
+    def __init__(self, vm, hypervisor):
+        self.vm = vm
+        self.hypervisor = hypervisor
+        self.ranks = []
+
+    def cmp(self, other):
+        assert self.vm == other.vm
+
+        # Compare the current preferences
+        rank_len = min(len(self.ranks), len(other.ranks))
+        ranks = self.ranks[:rank_len]
+        other_ranks = other.ranks[:rank_len]
+        if ranks < other_ranks:
+            return -1
+        if ranks > other_ranks:
+            return 1
+
+        # Check new preferences
+        for index in range(rank_len, len(HYPERVISOR_PREFERENCES)):
+            rank = self.get_rank(index)
+            other_rank = other.get_rank(index)
+            if rank < other_rank:
+                return -1
+            if rank > other_rank:
+                return 1
+
+        log.warning(
+            'Hypervisors "{}" and "{}" has exact same preference.'
+            .format(self.hypervisor.fqdn, other.hypervisor.fqdn)
+        )
+        return 0
+
+    def get_rank(self, index):
+        if len(self.ranks) == index:
+            preference = HYPERVISOR_PREFERENCES[index]
+            rank = preference(self.vm, self.hypervisor)
+            self.ranks.append(rank)
+        else:
+            rank = self.ranks[index]
+
+        return rank
+
+
+def get_hypervisors(vm, hypervisor_states=['online']):
+    filters = {
+        'servertype': 'hypervisor',
+        'environment': environ.get('IGVM_MODE', 'production'),
+        'vlan_networks': vm.dataset_obj['route_network'],
+        'state': Any(*hypervisor_states),
+    }
+
+    log.info('Evaluating hypervisors...')
+
+    return (r.hypervisor for r in sorted((
+        HypervisorRanking(vm, Hypervisor(o))
+        for o in Query(filters, HYPERVISOR_ATTRIBUTES)
+    ), key=cmp_to_key(HypervisorRanking.cmp), reverse=True))
```

Comment:
`key=cmp_to_key(HypervisorRanking.cmp),` is not the easiest construct to get.

Reply:
I will comment it.